Better Text Rendering for cairo / lua

simotek opened the pull request:
# Description
## Introduction
I have attempted to resolve #900 by creating a new lua binding specifically for doing font rendering in Cairo using Freetype and Harfbuzz, fontconfig is also used to locate the correct fonts. I have chosen to follow the approach of the `imlib2_helper` binding and not expose functions from the headers of each of these libraries but rather provide simple functions that "internally" use each of these libraries.

At the time of creating this PR I have no fixed thoughts on the API other then it works for my use case and tries to balance the simple act of rendering an English string  while also allowing complex options such as RTL text as requested in #75 as well as being able to center and right align text #940  and #1017.  So your thoughts on the API topic would be much appreciated, once the API is finalized I will add appropriate documentation. currently this is the last major step for be before being able to move to a lua only config.

## Current API Description (Happy to make changes here)
Loading a font from a file is a slow enough operation that we probably don't want to do it multiple times so I separated loading fonts and freeing memory at close into separate functions.

```
/* font is a string containing the name of the font to load */
/* this will be matched on a best effort basis and FontConfig will fall back to a similar font if the specified font isn't found */
FontData * cairo_text_hp_load_font(const char *font, int font_size);
void cairo_text_hp_destroy_font(FontData *font);
```

The next set of functions cover the "Simple" English use cases, I could have added a parameter for alignment but I was trying to keep the common Left case as simple as possible.

```
void cairo_text_hp_simple_show(cairo_t *cr, int x, int y, const char *text, FontData *font);
void cairo_text_hp_simple_show_center(cairo_t *cr, int x, int y, const char *text, FontData *font);
void cairo_text_hp_simple_show_right(cairo_t *cr, int x, int y, const char *text, FontData *font);
```
The next function is the "catch all" with the main implementation and allows tweeking of the internalization options such as text direction in Harfbuzz for people looking to use non latin languages.
```
/* 
 * Direction calls hb_direction_from_string example values are LTR and RTL
 *   https://harfbuzz.github.io/harfbuzz-hb-common.html#hb-direction-from-string
 * Script is an ISO 15924 4 character string, "Zyyy" can be used for "Common" and "Zinh"
 *  for "Inherited". 
 *  https://harfbuzz.github.io/harfbuzz-hb-common.html#hb-script-from-string
 * Language is a BCP 47 language tag. eg "en" or "en-US"
 */
void cairo_text_hp_intl_show(cairo_t *cr, int x, int y, cairo_text_alignment_t alignment, const char *text, 
                                                    FontData *font,  const char *direction, const char *script, const char *language);
```
Finally when doing layouting sometimes its useful to know how wide the final string will be so the following find the text width,  again there is a simple latin version and an internationalized version. In the future I'll also provide a Height function which will be important for supporting vertical text.

```
int cairo_text_hp_simple_text_width(const char *text, FontData *font);
int cairo_text_hp_intl_text_width(const char *text, FontData *font, const char *direction, const char *script, const char *language);
```

## Screenshot
In the following screenshot the "Cairo" text is using Cairo's built in basic text rendering, the "Conky" text is using Conky's standard text rendering and the "Freetype" strings are using this new API with Freetype and Harfbuzz rendering the text in Cairo.  All these cases are just using the default settings as found on openSUSE Tumbleweed.

![shot-2023-04-11_21-13-49](https://user-images.githubusercontent.com/3827011/231317226-b582b2b4-68e3-4061-a417-5f831869ba7e.jpg)

simotek committed:
lua text: Supporting infra

Add empty files and various build options for the lua text bindings
Adds requirements for Fontconfig, Freetype and Harfbuzz

simotek committed:
lua text: Supporting infra

Add empty files and various build options for the lua text bindings
Adds requirements for Fontconfig, Freetype and Harfbuzz

simotek committed:
lua text: initial implementation.

simotek committed:
lua text: initial implementation.

simotek committed:
Add the ability to center and right align text.

simotek committed:
lua text: Remove test code

I thought I already removed this and squashed the commit

brndnmtthws commented on .github/workflows/build-and-test-linux.yaml:
We could probably enable this in the macos build too, if you want to take a stab at that.

simotek replied:
Once I figure out why the ubuntu builds are failing i'll have a shot at enabling it.

simotek replied:
this seems to be working

brndnmtthws commented on lua/libcairo_text_helper.h:
The functions in here to be documented in `conky/doc/lua.yml`.

simotek replied:
I've had a shot at doing this in a way that hopefully makes sense, for some reason i couldn't generate docs locally but that doesn't seem related to any of my changes.

If your happy with that i'll probably do similar for the imlib2 bindings that also aren't currently documented.

simotek replied:
Also now done.

brndnmtthws reviewed (changes requested):
Looks great overall, just need to add documentation to https://github.com/brndnmtthws/conky/blob/main/doc/lua.yaml for the new functions.

simotek commented:
Thanks for the feedback, i'll update the docs and finish of implementing the Top to Bottom features then finish up a couple of configs to use them before sending through some updates. But its good to know i'm heading in the right direction.

moceap commented:
After trying this patches, bug #75 still exist!

![Screenshot_٢٠٢٣٠٧٠٤_٠٣٢٨٤١](https://github.com/brndnmtthws/conky/assets/4345211/48366e98-4278-4bde-8395-5aef228fab34)

simotek commented:
> After trying this patches, bug #75 still exist!
> 
> ![Screenshot_٢٠٢٣٠٧٠٤_٠٣٢٨٤١](https://user-images.githubusercontent.com/4345211/250696192-48366e98-4278-4bde-8395-5aef228fab34.png)

Yes that is currently to be expected, the WIP patch just had the LTR support so I could get feedback on the API, RTL is almost there I just need to finish writing the tests.

Also this change only works with the lua API, however it is the first step in my project to add a toolkit to make using the lua API much simpler. Although that is further from completion.

simotek commented:
@moceap Can you confirm that the following screenshot looks ok, it should be the days of the week.
![shot-2023-07-20_15-27-09](https://github.com/brndnmtthws/conky/assets/3827011/90f81563-a0ad-4391-a760-9cc6bba8957e)

I'll probably do just a little more testing of my actual real world usecases before sending through an updated pull request

simotek commented:
![shot-2023-07-20_15-51-45](https://github.com/brndnmtthws/conky/assets/3827011/150a7856-c95b-43e1-8ac4-682eff694b0e) 

This one uses Noto Sans Arabic instead.

moceap commented:
No it's not OK.

Letters in Arabic should be connected:

You write:

الإثنين
يوم الثلاثاء
الأربعاء
يوم الخميس
جمعة
السبت
الأحد

Look at them here in the browser, they are connected.

simotek commented:
> No it's not OK.
> 
> Letters in Arabic should be connected:
> 
> You write:
> 
> الإثنين يوم الثلاثاء الأربعاء يوم الخميس جمعة السبت الأحد
> 
> Look at them here in the browser, they are connected.

Thanks for the confirmation, that's what I thought i'll have to have a closer look at whats going wrong, the left to right seems to be ok.

Caellian commented on lua/libcairo_text_helper.h:
```diff
@@ -0,0 +1,281 @@
+  //alternatively you can use hb_buffer_set_unicode_funcs(buf, hb_glib_get_unicode_funcs());
+  hb_buffer_set_unicode_funcs(buf, hb_glib_get_unicode_funcs());
+  hb_buffer_set_direction(buf, text_direction); /* or LTR */
+  hb_buffer_set_script(buf, text_script); /* see hb-unicode.h */
```
Which `text_script` are you using to test out arabic? It should be "Arab" as defined in [`hb-common.h`](https://github.com/harfbuzz/harfbuzz/blob/fd5eb10d3b4cfee1bb10614ff6719378c949492e/src/hb-common.h#L526C17-L526C17). I'm asking because you're referencing `hb-unicode.h` which might be the reason why HarfBuzz is not shaping the arabic example properly.

Caellian commented on lua/libcairo_text_helper.h:
```diff
@@ -0,0 +1,281 @@
+  }
+  hb_script_t text_script = hb_script_from_string(script, -1);
+  if (text_script == HB_SCRIPT_UNKNOWN) {
+    text_script = HB_SCRIPT_COMMON;
```
Try setting this to `HB_SCRIPT_ARABIC` (string literal) instead of `HB_SCRIPT_COMMON` and see if it works (though it will break everything else).

Caellian commented on lua/libcairo_text_helper.h:
```diff
@@ -0,0 +1,281 @@
+  //alternatively you can use hb_buffer_set_unicode_funcs(buf, hb_glib_get_unicode_funcs());
+  hb_buffer_set_unicode_funcs(buf, hb_glib_get_unicode_funcs());
+  hb_buffer_set_direction(buf, text_direction); /* or LTR */
+  hb_buffer_set_script(buf, text_script); /* see hb-unicode.h */
```
This should be `hb-common.h` as well. Just pointing it out (so it's easier for others to pick up), not that these comments will affect anything.

Caellian commented on lua/libcairo_text_helper.h:
```diff
@@ -0,0 +1,281 @@
+    text_script = HB_SCRIPT_COMMON;
+  }
+    
+  hb_language_t text_language = hb_language_from_string (language, -1);
```
Is the language set to "ar"?

simotek replied:
Yep, the full setting was {font_family="Noto Sans Arabic", font_size=16, align=CAIRO_TEXT_ALIGN_LEFT, font_direction="RTL", font_script="HB_SCRIPT_ARABIC", font_language="ar"}

Caellian reviewed (commented):
Never used HarfBuzz, but from what I picked up googling it, the way you're using it seems fine.

Caellian commented on lua/libcairo_text_helper.h:
```diff
@@ -0,0 +1,281 @@
+  hb_buffer_destroy(buf);
+}
+
+void cairo_text_hp_simple_show(cairo_t *cr, int x, int y, const char *text, FontData *font)
```
Note that there's no need for these `cairo_text_hp_simple_show_*` convenience functions. toluapp supports default values and harfbuzz can infer defaults too.

You can make all of the properties optional by:
- adding a call to `hb_buffer_guess_segment_properties (hb_buffer);` in `cairo_text_hp_intl_show` after _specified properties_ are set.
- setting argument order to be: `language` (because it's inferred from system (least smart)), `script` (inferred from buffer contents), `direction` (inferred from script)
  - as documented [here](https://harfbuzz.github.io/harfbuzz-hb-buffer.html#hb-buffer-guess-segment-properties)
  - that order would minimize the need for supplying well inferred values to set the one without a good default
- specifying argument defaults to be `nullptr`/`NULL` in the header
  - and not calling `hb_*_from_string` if the value is null (i.e. empty string).

Caellian replied:
In the same line of reasoning I'd position cairo `alignment` argument after the `font` argument and set it to default to `CAIRO_TEXT_ALIGN_LEFT`.

Caellian replied:
Also, the binding function should be able to load and cache a font. The cache can leak memory (no need to call free) as it would be weird to manage such things from a lua script and fonts will likely be needed for each draw call.

So the signature of `cairo_text_hp_intl_show` in .pkg would finally look like: `void cairo_text_hp_intl_show(cairo_t *cr, int x, int y, const char *text, const char *font = NULL, int font_size = 12, cairo_text_alignment_t alignment = CAIRO_TEXT_ALIGN_LEFT, const char *language = NULL, const char *script = NULL, const char *direction = NULL)`.

simotek replied:
Sorry I didn't see this until now, I'll certainly swap to the idea of default arguments that seems much better. Leaking fonts seems reasonable to me as well so i'm happy to just add a cache.

simotek replied:
I am implementing most of this, with the exception of `language` which i'm going to default to english if not explicitly set. The reason being if I create a config / script for conky and want to distribute it, I don't want it to start breaking in random ways because other people are using a different language. Also much of conky's outputs such as process names CPU percentages etc work best in English. In this case if you are doing something that requires an explicit language you'll need to set it as well.

moceap commented:
@simotek Take a look https://github.com/HOST-Oman/libraqm, It may fast the process.

simotek commented:
> @simotek Take a look https://github.com/HOST-Oman/libraqm, It may fast the process.

One thing I noticed at https://github.com/HOST-Oman/libraqm/blob/9d7ca9d59dba38e777bdcd096870776dab5e6e66/src/raqm.c#L1793C1-L1793C48 is that for the RTL case they seem to have hardcoded a -1 into the calculation for character position, So I might have to have a play with that later. 

It seemed like hardcoding a -1 into the width of each character would solve the issue but I didn't want to just do it without seeing it documented, but the fact this other codebase is suggests its probably the right solution

simotek committed:
Commit the WIP so I can merge it

simotek committed:
Fix for #1698

This should go into its own PR but right now I need it for my system
to work.

simotek committed:
Fix comment hb-unicode.h should be hb-common.h

simotek committed:
Swap to using optional parameters

This replaces the old approach of using a series of function
declarations.

simotek committed:
Add Hack for RTL positioning

For some reason RTL positioning is wrong, it seems somewhat based
on font size 1/10th font size seems to get closer to correct so
run with that for now, i'll continue investigating a proper fix.

simotek committed:
Remove Top to Bottom Fixme's they are fixed

simotek committed:
Merge branch 'lua-text-helper2' into lua-text-helper3

simotek committed:
Some general cleanup post merge

simotek commented:
So far I've done all the interface changes, other then caching fonts, which is next on my list.

As for the RTL codepaths, subtracting 1/10th of the font size seems to get much closer then -1 especially for larger font sizes. Its not perfect for all fonts at all sizes though, but is significantly better, i'll follow this up with upstream closer

@moceap I presume this screenshot is getting much closer. At this point I have no idea if the text is actual words.

![shot-2024-02-15_11-53-40](https://github.com/brndnmtthws/conky/assets/3827011/c93be17a-7b15-4507-b2d2-b80a62d30b5c)

simotek committed:
Attempt to fix workflows by adding freetype dep

simotek committed:
Attempt to fix workflows by adding freetype dep

simotek committed:
Probably actually need freetype6

simotek committed:
Probably actually need freetype6

simotek committed:
Try again, I don't understand Ubuntu

simotek committed:
Try again, I don't understand Ubuntu

simotek committed:
Try something from google

simotek commented:
Also if anyone has any idea what Ubuntu packages are needed to get cmake to find freetype that would be much appreciated so that I don't need to spin up a VM and test.

simotek committed:
lua-text: Move to caching fonts in C

Fonts are now cached in C, given the number of fonts and sizes
I expect people will use, this implementation seems simpler then
implementing hashmaps or attempting to do bindings in C++

simotek committed:
Revert "Fix for #1698"

This reverts commit 5036cc4d09b267129351ad0d5485ca139aa98fd7.
Fixed better upstream now

simotek committed:
docs: tolua++ now works with more then lua5.1

On my machine i'm using it with lua5.3 and I see packages for
5.4 as well.

simotek committed:
lua-text: Add docs.

simotek committed:
lua-text: Fix build with

Caellian commented on doc/lua.yaml:
```diff
@@ -184,3 +184,44 @@ values:
+      | script              | A string containing a ISO 15924 script tag.                                          |
+      |                     | Default value: auto-detect from text                                                 |
+      | direction           | A string representing text direction eg `LTR`, `RTL` or `TTB`                        |
+      |                     | Default value: auto-detect from text otherwise `LTR`                                 |                                                               |
```
Extra pipe (`|`) at the end.

Caellian commented on doc/lua.yaml:
```diff
@@ -184,3 +184,44 @@ values:
+      | script              | A string containing a ISO 15924 script tag.                                          |
+      |                     | Default value: auto-detect from text                                                 |
+      | direction           | A string representing text direction eg `LTR`, `RTL` or `TTB`                        |
+      |                     | Default value: auto-detect from text otherwise `LTR`                                 |
```
Instead of putting default value into the next row, it might be better to put it in the same cell and use `<br/>` so that it looks like this:

| Argument | Description |
|-|-|
| language | A string containing a BCP 47 language tag. <br/>  Default value: `en` |

simotek replied:
This also should be fixed.

Caellian reviewed (commented):
Some inputs on tables in docs.

Caellian commented:
> Also if anyone has any idea what Ubuntu packages are needed to get cmake to find freetype that would be much appreciated so that I don't need to spin up a VM and test.

Should be `libfreetype6-dev`, see: https://askubuntu.com/questions/1347131/what-is-the-reason-of-cmake-error-when-using-freetype2-header
- CMake is crappy at locating dependencies sometimes (amongst a rather long list of other things), if you have that package installed, send the output of `apt-file list libfreetype6-dev` so we can patch the [Find script](https://cmake.org/cmake/help/latest/module/FindFreetype.html).

simotek committed:
Merge branch 'main' into lua-text-helper3

simotek committed:
lua-text use cairo_fill_preserve

This means its possible to add borders from lua after

simotek committed:
Doc: Implement feedback from #1501 for tables

simotek committed:
Merge branch 'main' into lua-text-helper3

simotek committed:
Remove code accidently re added in merge

simotek committed:
lua-text: Another shot at fixing workflows

This time add libfreetype-dev as well.

simotek committed:
cmake: try and use find_package for freetype

The old version worked on openSUSE but not ubuntu's CI, this
version also works on openSUSE lets see if it works in the CI

simotek commented:
> > Also if anyone has any idea what Ubuntu packages are needed to get cmake to find freetype that would be much appreciated so that I don't need to spin up a VM and test.
> 
> Should be `libfreetype6-dev`, see: https://askubuntu.com/questions/1347131/what-is-the-reason-of-cmake-error-when-using-freetype2-header
> 
>     * CMake is crappy at locating dependencies sometimes (amongst a rather long list of other things), if you have that package installed, send the output of `apt-file list libfreetype6-dev` so we can patch the [Find script](https://cmake.org/cmake/help/latest/module/FindFreetype.html).

I swapped it and the XFT implementation to using cmake's `find_package(Freetype REQUIRED)`, the old XFT implementation worked fine for me on openSUSE, but obviously it had issues on ubuntu

simotek commented:
I think this is finally getting close and i've finally got it doing everything I need from it.

All the text in the below screenshot is using the new conky lua API including the stiff with borders and Gradients etc. The conky memory leak is unrelated and fixed in #2083

![](http://www.enlightenment.org/ss/e-673abdb7381118.57957033.jpg)

brndnmtthws commented on Dockerfile:
```diff
@@ -40,11 +43,25 @@ RUN apt-get update \
   && apt-get clean \
   && rm -rf /var/lib/apt/lists/*
 
+# Compile CMake, we need the latest because the bug here (for armv7 builds):
```
Is adding this back intentional? It was removed in https://github.com/brndnmtthws/conky/commit/a533bcb8e41c57fdfa7a4aec7f3c2cd6c34734fe, so I'm wondering if this is just a merge issue.

brndnmtthws commented on cmake/ConkyBuildOptions.cmake:
```diff
@@ -26,15 +26,15 @@ if(NOT CMAKE_BUILD_TYPE)
       CACHE
-      STRING
-      "Choose the type of build, options are: None Debug Release RelWithDebInfo MinSizeRel."
+        STRING
```
It looks like the only thing that changed in this file is the formatting. I think it would be better to preserve the formatting here.

brndnmtthws commented on cmake/ConkyPlatformChecks.cmake:
```diff
@@ -499,32 +494,32 @@ set(conky_libs ${conky_libs} ${LUA_LIBRARIES})
 
-# Check for libraries used by Lua bindings
-if(BUILD_LUA_CAIRO)
+  # Check for libraries used by Lua bindings
```
Preserve formatting.

brndnmtthws commented on lua/CMakeLists.txt:
```diff
@@ -3,7 +3,7 @@
 # Please see COPYING for details
 #
-# Copyright (c) 2005-2024 Brenden Matthews, et. al. (see AUTHORS) All rights
+# Copyright (c) 2005-2021 Brenden Matthews, et. al. (see AUTHORS) All rights
```
Preserve this, along with the formatting.

brndnmtthws commented on lua/libcairo_text_helper.h:
```diff
@@ -0,0 +1,447 @@
+ *
+ * Please see COPYING for details
+ *
+ * Copyright (c) 2005-2021 Brenden Matthews, Philip Kovacs, et. al.
```
2024 (not a big deal, just noticed it).

brndnmtthws reviewed (changes requested):
The code looks pretty reasonable to me. A few minor nit picks (mostly formatting related). I'm not really familiar with the implementation details or harfbuzz, so I'll trust your judgement on that.

Not related to this PR, but we should probably separate the helper header declarations from the implementation code (putting the latter into separate files).

elParaguayo committed:
Fix background rendering for wayland output

Previously, the background was rendered with cairo's CLEAR operator
which just clears the surface and does not render any visible contents.

The changes here first clear the surface (with CAIRO_OPERATOR_CLEAR) and
then use (CAIRO_OPERATOR_OVER) to render the contents. The result is
that the background colour is rendered correctly, including with
semi-transparent colours.

simotek committed:
Merge branch 'main' into lua-text-helper3

simotek committed:
Fix an uneeded formatting change

simotek committed:
Fix formatting

simotek committed:
Bump Copyright year

simotek committed:
Use freetype2 and bump copyright

simotek committed:
Revert "Use freetype2 and bump copyright"

This reverts commit c2d2b987d91b5439c09a667a8fa52dc73a2848ef.

simotek committed:
Reverted wrong commit, fix memory leak again

simotek committed:
Don't need libs for freetype

simotek committed:
See if the CI likes this version better

simotek committed:
Dockerfile: fix regression in builder

simotek committed:
More whitespace fixes

simotek committed:
More whitespace

simotek committed:
Fixed an incorrect comment

simotek committed:
Fix alignment for TTB Text

simotek commented:
I will write a blog post to go with the release but here is a couple of simple and more complex examples if anyone wants to have a play with them.  I'm going to use it as part of a larger lua framework that's creating some of the above screenshots [https://github.com/simotek/conky-bubbles](https://github.com/simotek/conky-bubbles) which will also get a blog post and probably a conference talk or two.

![shot-2025-10-29_17-41-29](https://github.com/user-attachments/assets/6c778672-ccb1-4d85-b33d-9e43a30a6907)

```
function conky_text_example()
    -- The script may be called before the window is created.
    if conky_window == nil then
        return nil
    end

    -- Create Surface and cairo object for drawing
    local cs = cairo_xlib_surface_create(conky_window.display,
                                         conky_window.drawable,
                                         conky_window.visual,
                                         conky_window.text_width,
                                         conky_window.text_height)
    ---
    -- Simple example displaying a date
    ---
    local cr = cairo_create(cs)
    local date_time = os.date("%Y-%m-%d %H:%M:%S")
    local font_family = "SUSE Thin"
    local font_size = 16
    local w, h = cairo_text_hp_text_size(date_time, font_family, font_size)
    local x = (conky_window.text_width - w)/2
    cairo_set_source_rgba(cr, 1.0,1.0,1.0,1.0)
    cairo_set_line_width(cr, 0.2)
    cairo_text_hp_show(cr, x, 10, date_time, font_family, font_size)

    cairo_stroke(cr)

    ---
    -- Complex fancy example
    ---

    -- Create a mesh with a background
    local title_gradient = cairo_pattern_create_mesh()
    cairo_mesh_pattern_begin_patch(title_gradient)
    cairo_mesh_pattern_line_to(title_gradient, 00, 10)
    cairo_mesh_pattern_line_to(title_gradient, 160, 5)
    cairo_mesh_pattern_line_to(title_gradient, 150, 55)
    cairo_mesh_pattern_line_to(title_gradient, -10, 65)
    cairo_mesh_pattern_line_to(title_gradient, 00, 10)

    -- Unfortunately Cairo uses values between 0.0 and 1.0 for rgb values
    cairo_mesh_pattern_set_corner_color_rgba(title_gradient, 0, 0.20, 0.59, 0.84, 1.00)
    cairo_mesh_pattern_set_corner_color_rgba(title_gradient, 1, 0.55, 0.26,	0.67, 1.00)
    cairo_mesh_pattern_set_corner_color_rgba(title_gradient, 2, 0.20, 0.59, 0.84, 1.00)
    cairo_mesh_pattern_set_corner_color_rgba(title_gradient, 3, 0.55, 0.18, 0.14, 1.00)
    cairo_mesh_pattern_end_patch(title_gradient)
    cairo_set_source(cr, title_gradient)

    -- Render the text
    cairo_text_hp_show(cr, 10, 40, "TESTING 1234", "Sixteen", 16)
    -- Border Line Width and Color
    cairo_set_line_width(cr, 1.2)
    cairo_set_source_rgba(cr, 0.00, 0.32, 0.52, 1.00)
    cairo_stroke(cr)

    ---
    -- Some Right to left and Top to bottom text
    ---
    cairo_set_source_rgba(cr, 1.0,1.0,1.0,1.0)
    
    local w, h = cairo_text_hp_text_size("السبت", "Noto Sans Arabic", 16, 
                                         CAIRO_TEXT_ALIGN_RIGHT, "ar", "HB_SCRIPT_ARABIC", "RTL")
    
    cairo_text_hp_show(cr, conky_window.text_width - w -25, 60, "السبت", "Noto Sans Arabic", 16,
                       CAIRO_TEXT_ALIGN_RIGHT, "ar", "HB_SCRIPT_ARABIC", "RTL")
    -- Use the same formatting for this top to bottom text
    cairo_text_hp_show(cr, 10, 70, "這是一些中文", "Source Han Sans TW", 16, 
                       CAIRO_TEXT_ALIGN_LEFT, "ch", "HB_SCRIPT_HAN", "TTB")

    -- Cleanup Memory usage
    cairo_destroy(cr)
    cairo_surface_destroy(cs)
end
```

Caellian committed:
Fix lvalue member reference in geometry

Also removed C++ 14 detection from Vc, now forcing C++14 from Vc build as C++17 is the minimum required version for Conky anyway.

Signed-off-by: Tin Švagelj <tin.svagelj@live.com>

mfluhr committed:
Fixed typo inside the hwmon parse error message

brndnmtthws committed:
chore: only run CI for macos-latest

pkkid committed:
Fix mouse button events for owned windows with XInput2; Fixes #2047

brndnmtthws committed:
fix(proc): harden /proc helpers and add tests

- reset cmdline_to_pid pathstream each iteration to avoid stale paths
- parse /proc/[pid]/stat by skipping comm field so spaces are supported
- prevent pid_thread_list writes from overflowing small buffers
- add Linux tests for cmdline_to_pid, pid_time, and pid_thread_list
  safety

brndnmtthws committed:
fix(proc): harden /proc helpers

brndnmtthws committed:
fix: ignore non-pid entries in cmdline_to_pid

brndnmtthws committed:
fix: parse proc stat without suppressed length modifiers

brndnmtthws committed:
test(proc): stabilize pid_environ lookup

brndnmtthws committed:
test(proc): tolerate vmrss drift

brndnmtthws committed:
test(proc): freeze child for vm status

brndnmtthws committed:
fix(darwin): correct top cpu percentage math

Normalize Darwin process CPU times to centiseconds using
mach_timebase_info rather than assuming nanoseconds, and avoid reusing
aggregate tick totals across updates. Align total CPU normalization with
expected per-core vs aggregate behavior and guard zero deltas. Add
helper module for Darwin top calculations and unit tests covering sample
reset, delta init, and timebase conversion.

Fixes #2276.

simotek commented:
@brndnmtthws I think this is ready for final review again

brndnmtthws committed:
fix(lua): initialize default harfbuzz language

Ensure fallback language assignment when Lua callers omit "language".

Applies to both cairo_text_hp_show and cairo_text_hp_text_size to
prevent uninitialized hb_language_t usage.

brndnmtthws committed:
Merge branch 'main' into simotek-lua-text-helper3
